src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+    {
+        private string _processorId;
+
+        private List<int> _ownedPartitions;
+
+        private BlobContainerClient _containerClient;
+
+        private int _totalProcessorsRunning = 0;
+
+        private string[] _eventHubPartitions;
+
+        private EventBatchingService _eventBatchingService;
+
+        private StorageCheckpointClient _checkpointClient;
+
+        private ITelemetryLogger _logger;
+
+        private PartitionLockingBackgroundServiceOptions _partitionLockingOptions;
+
+        private EventHubClientOptions _eventHubClientOptions;
+
+        public PartitionLockingBackgroundService(
+            PartitionLockingBackgroundServiceOptions partitionLockingOptions,
+            EventHubClientOptions eventHubClientOptions,
+            EventBatchingService eventBatchingService,
+            StorageCheckpointClient checkpointClient,
+            ITelemetryLogger logger)
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
+            _partitionLockingOptions = partitionLockingOptions;
+            _eventHubClientOptions = eventHubClientOptions;
+            _eventBatchingService = eventBatchingService;
+            _checkpointClient = checkpointClient;
+            _logger = logger;
+        }
+
+        private void CreateContainerClient()
+        {
+            var tokenCredential = EnsureArg.IsNotNull(_partitionLockingOptions.StorageTokenCredential, nameof(_partitionLockingOptions.StorageTokenCredential));
+            _containerClient = new BlobContainerClient(_partitionLockingOptions.BlobContainerUri, tokenCredential.GetCredential());
+            _containerClient.CreateIfNotExists();
+        }
+
+        private async Task RegisterProcessor(CancellationToken ct)
+        {
+            // write process id to storage
+            _logger.LogTrace($"Processor Id: {_processorId}");
+            BlobClient blobClient = _containerClient.GetBlobClient($"activepods/{_processorId}");
+
+            var activeUntil = new Dictionary<string, string>()
+            {
+                { "activeuntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+            };
+
+            BlobUploadOptions options = new BlobUploadOptions() { Metadata = activeUntil };
+            await blobClient.UploadAsync(BinaryData.FromString(string.Empty), options, ct);
+
+            _logger.LogTrace($"Registered processor {_processorId}");
+        }
+
+        private async Task RegisterProcessorLoop(CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                await RegisterProcessor(ct);
+                await Task.Delay(15000, ct);
+            }
+        }
+
+        private async Task<string[]> GetPartitionsIdsForEventHub(CancellationToken ct)
+        {
+            await using (var consumer = new EventHubConsumerClient(
```

Comment:
Does a new instance of `EventHubConsumerClient` need to be created each time? This seems like a significant perf consideration. It looks like none of the values in the constructor ever change, so is there a reason we can't abstract this behind an interface and inject a singleton into this class?

Reply:
that makes sense. it is injected now as a singleton